filament/backend/include/private/backend/HandleAllocator.h

```diff
@@ -179,21 +179,40 @@ class HandleAllocator {
                         << "use-after-free of Handle with id=" << handle.getId()
                         << ", tag=" << getHandleTag(handle.getId()).c_str_safe();
             }
+        } else {
+            // check for heap handle use-after-free
+            if (UTILS_UNLIKELY(!mUseAfterFreeCheckDisabled)) {
+                uint8_t const index = (handle.getId() & HANDLE_INDEX_MASK);
+                // if we've already handed out this handle index before, it's definitely a
+                // use-after-free, otherwise it's probably just a corrupted handle
+                if (index < mId) {
+                    FILAMENT_CHECK_POSTCONDITION(p != nullptr)
+                            << "use-after-free of heap Handle with id=" << handle.getId()
+                            << ", tag=" << getHandleTag(handle.getId()).c_str_safe();
+                } else {
+                    FILAMENT_CHECK_POSTCONDITION(p != nullptr)
+                            << "corrupted heap Handle with id=" << handle.getId()
+                            << ", tag=" << getHandleTag(handle.getId()).c_str_safe();
+                }
+            }
         }
 
         return static_cast<Dp>(p);
     }
 
     template<typename B>
     bool is_valid(Handle<B>& handle) {
-        if (handle && isPoolHandle(handle.getId())) {
+        if (!handle) {
```

Comment:
I think a null handle is considered valid (at least it was before). Maybe we should keep that behavior for now?

Reply:
Hm, it seems strange to me that a null handle is considered valid, but you're right, that's the way it was before.